src/python/pants/engine/internals/native.py

```diff
@@ -34,35 +37,78 @@
 logger = logging.getLogger(__name__)
 
 
+class IntrinsicError(ValueError):
+    """Exceptions raised for failures within intrinsic methods implemented in Rust."""
+
+
+class IncorrectProductError(TypeError):
+    """Exceptions raised when a rule's return value doesn't match its declared type."""
+
+
 class Externs:
     """Methods exposed from Python to Rust.
 
     TODO: These could be implemented in Rust in `externs.rs` via the cpython API.
     """
 
-    def __init__(self, lib):
+    def __init__(self, lib: ModuleType):
         self.lib = lib
 
+    # NB: This flag is exposed for testing error handling in CFFI methods. This should never be set
+    # in normal pants usage.
     _do_raise_keyboardinterrupt = bool(os.environ.get("_RAISE_KEYBOARDINTERRUPT_IN_EXTERNS", False))
 
-    def is_union(self, input_type):
+    def is_union(self, input_type: Type) -> bool:
         """Return whether or not a type is a member of a union."""
         # NB: This check is exposed for testing error handling in CFFI methods. This code path should
         # never be active in normal pants usage.
         return union.is_instance(input_type)
 
-    def create_exception(self, msg):
-        """Given a utf8 message string, create an Exception object."""
-        return Exception(msg)
+    def create_value_error(self, msg: str) -> IntrinsicError:
+        """Given a utf8 message string, create an exception signalling a value error."""
+        return IntrinsicError(msg)
+
+    def create_type_error(self, msg: str) -> IncorrectProductError:
+        """Given a utf8 message string, create an exception signalling a type error."""
+        return IncorrectProductError(msg)
```

Comment:
The convention for "calling a constructor" has moved toward declaring the type on `(Py)Types` (https://github.com/pantsbuild/pants/blob/83ed9f6cc1ea1f9de4b8eb9de10760474d792647/src/python/pants/engine/internals/scheduler.py#L132-L152 and https://github.com/pantsbuild/pants/blob/83ed9f6cc1ea1f9de4b8eb9de10760474d792647/src/rust/engine/src/externs/interface.rs#L396-L447) and then calling it via `unsafe_call`: eg: https://github.com/pantsbuild/pants/blob/83ed9f6cc1ea1f9de4b8eb9de10760474d792647/src/rust/engine/src/intrinsics.rs#L193-L196

Reply:
Going to try to make this change!

Reply:
Thanks!